src/mo_values/value.ml

```diff
@@ -180,75 +181,98 @@ let comma ppf () = fprintf ppf ",@ "
 
 let semi ppf () = fprintf ppf ";@ "
 
-let rec pp_val_nullary d ppf = function
-  | Null -> pr ppf "null"
-  | Bool b -> pr ppf (if b then "true" else "false")
-  | Int n when Int.(ge n zero) -> pr ppf (Int.to_pretty_string n)
-  | Int8 n when Int_8.(n = zero) -> pr ppf (Int_8.to_pretty_string n)
-  | Int16 n when Int_16.(n = zero) -> pr ppf (Int_16.to_pretty_string n)
-  | Int32 n when Int_32.(n = zero) -> pr ppf (Int_32.to_pretty_string n)
-  | Int64 n when Int_64.(n = zero) -> pr ppf (Int_64.to_pretty_string n)
-  | Nat8 n -> pr ppf (Nat8.to_pretty_string n)
-  | Nat16 n -> pr ppf (Nat16.to_pretty_string n)
-  | Nat32 n -> pr ppf (Nat32.to_pretty_string n)
-  | Nat64 n -> pr ppf (Nat64.to_pretty_string n)
-  | Float f -> pr ppf (Float.to_pretty_string f)
-  | Char c ->  pr ppf (string_of_string '\'' [c] '\'')
-  | Text t -> pr ppf (string_of_string '\"' (Lib.Utf8.decode t) '\"')
-  | Blob b -> pr ppf ("\"" ^ Blob.escape b ^ "\"")
-  | Tup vs ->
+let rec pp_val_nullary d ppf (t, v : T.typ * value) =
+  let t = T.normalize t in
+  match t, v with
+  | _, Null -> pr ppf "null"
+  | _, Bool b -> pr ppf (if b then "true" else "false")
+  | _, Int n when Int.(ge n zero) -> pr ppf (Int.to_pretty_string n)
+  | _, Int8 n when Int_8.(n = zero) -> pr ppf (Int_8.to_pretty_string n)
+  | _, Int16 n when Int_16.(n = zero) -> pr ppf (Int_16.to_pretty_string n)
+  | _, Int32 n when Int_32.(n = zero) -> pr ppf (Int_32.to_pretty_string n)
+  | _, Int64 n when Int_64.(n = zero) -> pr ppf (Int_64.to_pretty_string n)
+  | _, Nat8 n -> pr ppf (Nat8.to_pretty_string n)
+  | _, Nat16 n -> pr ppf (Nat16.to_pretty_string n)
+  | _, Nat32 n -> pr ppf (Nat32.to_pretty_string n)
+  | _, Nat64 n -> pr ppf (Nat64.to_pretty_string n)
+  | _, Float f -> pr ppf (Float.to_pretty_string f)
+  | _, Char c ->  pr ppf (string_of_string '\'' [c] '\'')
+  | _, Text t -> pr ppf (string_of_string '\"' (Lib.Utf8.decode t) '\"')
+  | T.Obj (T.Actor, _), Blob b ->
+    pr ppf (string_of_string '`' (Lib.Utf8.decode (Ic.Url.encode_principal b)) '`')
+  | _, Blob b -> pr ppf ("\"" ^ Blob.escape b ^ "\"")
+  | t, Tup vs ->
+    let list = match t with
+    | T.Tup ts -> List.combine ts vs
+    | _ -> List.map (fun v -> (T.Any, v)) vs in
     fprintf ppf "@[<1>(%a%s)@]"
-      (pp_print_list ~pp_sep:comma (pp_val d)) vs
+      (pp_print_list ~pp_sep:comma (pp_val d)) list
       (if List.length vs = 1 then "," else "")
-  | Obj ve ->
+  | t, Obj ve ->
+    let fs = match t with  T.Obj (_, fs) -> fs | _ -> [] in
     if d = 0 then pr ppf "{...}" else
     fprintf ppf "@[<hv 2>{@;<0 0>%a@;<0 -2>}@]"
-      (pp_print_list ~pp_sep:semi (pp_field d)) (Env.bindings ve)
-  | Array a ->
+      (pp_print_list ~pp_sep:semi (pp_field d)) (List.map (fun (lab, v) ->
+          let t = (List.find_map (fun T.{ lab = lab'; typ; _ } -> if lab = lab' then Some typ else None)) fs in
```

Comment:
```suggestion
          let t = Type.lookup_val_field_opt lab fs in
```
Lookup_val_field_opt correctly distinguishes between type fields and value fields -these actually share the same namespace. I think your code could be confused by `object { type x = int; val  x = 1}`.

Reply:
I guess this also raises the question of printing out type fields or suppressing them. Before, when we only had the value to hand, we could only print out the value fields. But now that we have the type available, we can print out both.